crates/consensus/src/transaction/envelope.rs

```diff
@@ -95,64 +103,90 @@ impl TryFrom<u8> for TxType {
 #[non_exhaustive]
 pub enum TxEnvelope {
     /// An untagged [`TxLegacy`].
-    Legacy(Signed<TxLegacy>),
+    Legacy(Enveloped<TxLegacy>),
     /// A [`TxEip2930`] tagged with type 1.
-    Eip2930(Signed<TxEip2930>),
+    Eip2930(Enveloped<TxEip2930>),
     /// A [`TxEip1559`] tagged with type 2.
-    Eip1559(Signed<TxEip1559>),
+    Eip1559(Enveloped<TxEip1559>),
     /// A TxEip4844 tagged with type 3.
     /// An EIP-4844 transaction has two network representations:
     /// 1 - The transaction itself, which is a regular RLP-encoded transaction and used to retrieve
     /// historical transactions..
     ///
     /// 2 - The transaction with a sidecar, which is the form used to
     /// send transactions to the network.
-    Eip4844(Signed<TxEip4844Variant>),
+    Eip4844(Enveloped<TxEip4844Variant>),
     /// A [`TxEip7702`] tagged with type 4.
-    Eip7702(Signed<TxEip7702>),
+    Eip7702(Enveloped<TxEip7702>),
 }
 
 impl From<Signed<TxLegacy>> for TxEnvelope {
     fn from(v: Signed<TxLegacy>) -> Self {
-        Self::Legacy(v)
+        let mut buf = Vec::with_capacity(v.tx().encoded_len_with_signature(v.signature()));
+        v.tx().encode_with_signature_fields(v.signature(), &mut buf);
```

Comment:
is there no other way to get the tx hash?

if the only solution is to go through the envelope type this will be very inconvenient

Reply:
c.f. #1483. Technically it's legal for networks to make envelopes that do the following:
- use the same inner type, but a different serialization (e.g. a TxLegacy serialized borsht instead of RLP)
- use a different tx flag for the same inner type (e.g. a `TxEnvelope` containing `TxEip1559` at `23` instead of `2`
- use the same inner type at multiple tx flags (e.g. a `TxEnvelope` containing `TxEip1559` at both `2` and `23`)

So there is no way to correctly get the hash of a transaction without wrapping it in the specific network's envelope first

Reply:
all of those usecases can be solved by introducing a separate inner transaction type (e.g. network can provide its own `TxLegacy`)

right now it makes total sense for separate transaction types to know their full encoding because we don't yet have a case of network rolling out encoding of transaction conflicting with existing tx types. also given that most of the networks target for compatibility with ethereum I'd consider such case very unlikely

Reply:
I also consider it unlikely. But it's still the only correct abstraction of an EIP-2718 envelope if we want our code to be reusable.

> because the type id is even part of the eip itself

This is a fair reading of the spec. And @klkvr is right that people downstream can just duplicate 500 lines of code per tx type to change a single constant. However, it costs us very little to have the envelope drive serialization and hashing

For another interesting example, consider the extension sentinel value in eip-2718 `0xff` 🙃  This is also unlikely ofc, but tx-association falls down in extended envelopes